### PR TITLE
Implement more callbacks

### LINKED ISSUE #94: Generalize on_best to on_outcome
ALNS currently supports a callback function to be called when ALNS finds a new global best solution state:

https://github.com/N-Wouda/ALNS/blob/55e5cfc8b77609b825512e4622478f48f1102183/alns/ALNS.py#L254-L267

Since we use the notion of outcomes in ALNS, we can generalize this hook to work for other outcomes such as better, accepted, and rejected as well. 

A simple proposal is to replace `on_best` with a new method:
```
def on_best(self, func: _OperatorType, outcome: Outcome):
     self._on_outcome[outcome] = func
```

I am not aware of any papers that do this. Maybe it's just a redundant addition.

### PR DESCRIPTION
**Pull request by N-Wouda**
Closes #94. This PR implements callbacks for each possible candidate solution evaluation.

Callbacks have the same signature as regular operators. They should thus return a new solution. This solution is then evaluated again.

### CODE REVIEW

**Comment by N-Wouda** [comment]
@leonlan this is ready for review. Adds a new `Outcome` enum, and new callbacks for the different outcomes.

File: alns/ALNS.py [hunk 1]
```diff
@@ -246,18 +259,41 @@ def _eval_cand(
             A tuple of the best and current solution, along with the weight
             index.
         """
-        w_idx = _REJECT
+        outcome = self._determine_outcome(accept, best, curr, cand)
+        func = self._on_outcome.get(outcome)
+
+        if callable(func):
+            cand = func(cand, self._rnd_state, **kwargs)
+            outcome = self._determine_outcome(accept, best, curr, cand)
```
Comment (leonlan) [inline alns/ALNS.py:267]: Issue: Whenever a callback function is used in this implementation, `accept` is called twice and hence also is updated twice. This is problematic as the thresholds decrease twice as fast, which is undesired behavior.

I propose the convention that the callbacks are applied to the _new_ `current` solution. 
- We evaluate the candidate as we did before.
- Based on the outcome, we apply the corresponding callback if available.
    - For best, better, accepted, the new current solution is the candidate solution. The callback is applied to this candidate solution.
    - For rejected, the new current solution is the old current solution. The callback is applied to the old current solution.
- Check (again) if the post-callback solution is a new best or not.

Comment (N-Wouda) [reply alns/ALNS.py:267]: I get the issue, but I'm not 100% sure about the proposed solution. Particularly, I dislike that `on_reject` would never see the rejected candidate. That defeats some of the purpose.

I will think a bit about alternatives.

Comment (leonlan) [reply alns/ALNS.py]: Yes this works for me!

**Comment by N-Wouda** [comment]
I'm merging this once the CI passes

**Review by leonlan — APPROVED** [review]
LGTM.

**Pull request merged**